src/e3/collection/dag.py

```diff
@@ -506,25 +511,26 @@ def check(self) -> None:
         else:
             self.__has_cycle = False
 
-    def get_closure(self, vertex_id: VertexID) -> FrozenSet[VertexID]:
+    def get_closure(self, vertex_id: VertexID) -> Set[VertexID]:
         """Retrieve closure of predecessors for a vertex.
 
         :param vertex_id: the vertex to inspect
         :return: a set of vertex_id
         """
-        visited: Set[VertexID] = set()
-        closure = self.get_predecessors(vertex_id)
-        closure_len = len(closure)
+        closure: Set[VertexID] = set()
+        to_visit = set(self.get_predecessors(vertex_id))
 
         while True:
-            for n in closure - visited:
-                visited.add(n)
+            next_visit: Set[VertexID] = set()
+            for n in to_visit:
+                closure.add(n)
+                next_visit |= self.get_predecessors(n)
 
-                closure |= self.get_predecessors(n)
+            to_visit = next_visit - closure
 
-            if len(closure) == closure_len:
+            if len(to_visit) == 0:
                 break
```

Comment:
It's possible to remove these lines and add it as the condition of the `while` loop

Reply:
Sorry miss your message. Indeed the code can be refactored. Missed that